File: ArticlesI24OBJC/ArticlesI24OBJC/MainTableViewController.m
//
//  MainTableViewController.m
//  ArticlesI24OBJC
//
//  Created by Menahem Barouk on 12/12/2019.
//  Copyright © 2019 Gini-Apps. All rights reserved.
//

#import "MainTableViewController.h"
#import "Article.h"
#import "ArticlesI24OBJC-Swift.h"

@interface MainTableViewController ()

@property (weak, nonatomic) IBOutlet UIBarButtonItem *showArticlesButton;
@property (weak, nonatomic) IBOutlet UISwitch* arabicSwitch;
@property (weak, nonatomic) IBOutlet UISegmentedControl* localeControls;
@property (weak, nonatomic) IBOutlet UIStepper* indexStepper;
@property (weak, nonatomic) IBOutlet UILabel* indexLabel;

@property (nonatomic) NSArray<Article *> *enArticles;
@property (nonatomic) NSArray<Article *> *arArticles;

@end

@implementation MainTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.arabicSwitch sendActionsForControlEvents:UIControlEventValueChanged];
    [self.localeControls sendActionsForControlEvents:UIControlEventValueChanged];
    [self.indexStepper sendActionsForControlEvents:UIControlEventValueChanged];
    [self fetchArticles];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self.navigationController setNavigationBarHidden:NO animated:animated];
}

- (void)fetchArticles
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        
        NSString *articleENPath = [NSBundle.mainBundle pathForResource:@"articlesEN" ofType:@"json"];
        NSString* jsonString = [[NSString alloc] initWithContentsOfFile:articleENPath encoding:NSUTF8StringEncoding error:nil];

        NSData* jsonData = [jsonString dataUsingEncoding:NSUTF8StringEncoding];

        NSArray *articlesENArr = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:nil];
        self.enArticles = [Article objectsWithDictionaries:articlesENArr];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            self.showArticlesButton.enabled = YES;
        });
    });
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        
        NSString *articleARURL = [NSBundle.mainBundle pathForResource:@"articlesAR" ofType:@"json"];
        NSString* jsonString = [[NSString alloc] initWithContentsOfFile:articleARURL encoding:NSUTF8StringEncoding error:nil];

        NSData* jsonData = [jsonString dataUsingEncoding:NSUTF8StringEncoding];

        
        NSArray *articleArARURL = [NSJSONSerialization JSONObjectWithData:jsonData options:0 error:nil];
        self.arArticles = [Article objectsWithDictionaries:articleArARURL];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            self.arabicSwitch.enabled = YES;
        });
    });
}

- (IBAction)arabicSwitchValueChanged:(UISwitch *)sender
{
    [VersionManager.sharedInstance updateLanguage:sender.isOn ? VersionManagerLanguageAr : VersionManagerLanguageEn];
}

- (IBAction)localeControlsValueChanged:(UISegmentedControl *)sender
{
    
}

- (IBAction)indexStepperValueChanged:(UIStepper *)sender
{
    self.indexLabel.text = [NSString stringWithFormat:@"Indexed at: %ld", (long)sender.value];
}

- (IBAction)showArticlesAction:(id)sender
{
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Articles" bundle:nil];
    ArticlesPageViewController *articlesPageViewController = [storyboard instantiateInitialViewController];
    [articlesPageViewController bindData:VersionManager.sharedInstance.isArabic ? self.arArticles : self.enArticles at:(NSInteger)self.indexStepper.value];
    [self showViewController:articlesPageViewController sender:self];
}

@end

File: ArticleI24NEWS/MainTableViewController.swift
//
//  MainTableViewController.swift
//  ArticleI24NEWS
//
//  Created by Menahem Barouk on 05/12/2019.
//  Copyright © 2019 Gini-Apps. All rights reserved.
//

import UIKit
import ProcedureKit

class MainTableViewController: UITableViewController {
    
    @IBOutlet weak var showArticlesButton: UIBarButtonItem!
    @IBOutlet weak var arabicSwitch: UISwitch!
    @IBOutlet weak var localeControls: UISegmentedControl!
    @IBOutlet weak var indexStepper: UIStepper!
    @IBOutlet weak var indexLabel: UILabel!
    
    var enArticles: [Article]!
    var arArticles: [Article]!
    let queue = ProcedureQueue()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        arabicSwitch.sendActions(for: .valueChanged)
        localeControls.sendActions(for: .valueChanged)
        indexStepper.sendActions(for: .valueChanged)
        fetchArticles()
    }
    
    override func viewWillAppear(_ animated: Bool)
    {
        super.viewWillAppear(animated)
        navigationController?.setNavigationBarHidden(false, animated: animated)
    }
    
    func fetchArticles()
    {
        let articleENURL = Bundle.main.url(forResource: "articlesEN", withExtension: "json")!
        let bringLocaleENArticlesOperation = BringJSONDataLocaleFileOperation(url: articleENURL, outputType: [Article].self)
        
        bringLocaleENArticlesOperation.addDidFinishBlockObserver(synchronizedWith: DispatchQueue.main) { [weak self] (networkOpration, error) in
            guard let strongSelf = self else { return }
            
            guard let articles = networkOpration.output.value?.value else { return }
            
            strongSelf.enArticles = articles
            
            if strongSelf.queue.operationCount == 0
            {
                strongSelf.showArticlesButton.isEnabled = true
            }
        }
        
        let articleARURL = Bundle.main.url(forResource: "articlesAR", withExtension: "json")!
        let bringLocaleARArticlesOperation = BringJSONDataLocaleFileOperation(url: articleARURL, outputType: [Article].self)
        
        bringLocaleARArticlesOperation.addDidFinishBlockObserver(synchronizedWith: DispatchQueue.main) { [weak self] (networkOpration, error) in
            guard let strongSelf = self else { return }
            
            guard let articles = networkOpration.output.value?.value else { return }
            
            strongSelf.arArticles = articles
            
            if strongSelf.queue.operationCount == 0
            {
                strongSelf.showArticlesButton.isEnabled = true
            }
        }
        
        queue.addOperations(bringLocaleENArticlesOperation, bringLocaleARArticlesOperation)
    }

    @IBAction func arabicSwitchValueChanged(_ sender: UISwitch) {
        VersionManager.shared.isArabic = sender.isOn
    }
    
    @IBAction func localeControlsValueChanged(_ sender: UISegmentedControl) {
        let identifiers = ["en_UK","fr_FR","ar_SA"]
        let identifier = identifiers[sender.selectedSegmentIndex]
        VersionManager.shared.locale = Locale(identifier: identifier)
    }
    
    @IBAction func indexStepperValueChanged(_ sender: UIStepper)
    {
        indexLabel.text = "Indexed at: \(Int(sender.value))"
    }
    
    @IBAction func showArticlesAction(_ sender: Any) {
        guard let articles = VersionManager.shared.isArabic ? arArticles : enArticles else { return }
        
        let storyboard = UIStoryboard(name: "Articles", bundle: nil)
        let articlesPageViewController = storyboard.instantiateInitialViewController() as! ArticlesPageViewController
        articlesPageViewController.bindData(articles, at: Int(indexStepper.value))
        show(articlesPageViewController, sender: self)
    }
}
